File: Blind 75/07 - Trees/100-same-tree.py
'''
Given the roots of two binary trees p and q, write a function to check if they are the same or not.

Two binary trees are considered the same if they are structurally identical, and the nodes have the same value.
'''

from collections import deque
from time import time
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        if not p and not q:
            return True
        if (not p and q) or (p and not q):
            return False

        q1, q2 = deque([p]), deque([q])
        while q1 and q2:
            node1, node2 = q1.popleft(), q2.popleft()

            if (
                node1.val != node2.val
                or (not node1.left and node2.left)
                or (node1.left and not node2.left)
                or (not node1.right and node2.right)
                or (node1.right and not node2.right)
            ):
                return False

            if node1.left and node2.left:
                q1.append(node1.left)
                q2.append(node2.left)
            if node1.right and node2.right:
                q1.append(node1.right)
                q2.append(node2.right)

        return True

    def reference(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        if not p and not q:
            return True
        if not p or not q or p.val != q.val:
            return False

        return self.isSameTree(p.left, q.left) and self.isSameTree(
            p.right, q.right
        )

    def quantify(self, test_cases, runs=50000):
        sol_start = time()
        for i in range(runs):
            for case in test_cases:
                if i == 0:
                    print(self.isSameTree(*case))
                else:
                    self.isSameTree(*case)
        print(f'Runtime for our solution: {time() - sol_start}\n')

        ref_start = time()
        for i in range(0, runs):
            for case in test_cases:
                if i == 0:
                    print(self.reference(*case))
                else:
                    self.reference(*case)
        print(f'Runtime for reference: {time() - ref_start}')


if __name__ == '__main__':
    test = Solution()
    test_cases = [
        (
            TreeNode(1, TreeNode(2), TreeNode(3)),
            TreeNode(1, TreeNode(2), TreeNode(3)),
        ),
        (TreeNode(1, TreeNode(2)), TreeNode(1, None, TreeNode(2))),
        (
            TreeNode(1, TreeNode(2), TreeNode(1)),
            TreeNode(1, TreeNode(1), TreeNode(2)),
        ),
    ]
    test.quantify(test_cases)

File: Blind 75/07 - Trees/JavaScript/100-same-tree.js
const { performance } = require('perf_hooks');

class TreeNode {
  constructor(val, left, right) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

class Solution {
  method(p, q) {
    const dfs = (p, q) => {
      if (!p && !q) return true;
      if (!p || !q || p.val !== q.val) return false;

      return dfs(p.left, q.left) && dfs(p.right, q.right);
    };

    return dfs(p, q);
  }

  reference(p, q) {
    const isBaseCase = !(p || q);
    if (isBaseCase) return true;

    const isBalanced = p && q;
    if (!isBalanced) return false;

    const isSame = p.val === q.val;
    if (!isSame) return false;

    return this.dfs(p, q);
  }

  dfs = (p, q) => {
    const left = this.reference(p.left, q.left);
    const right = this.reference(p.right, q.right);

    return left && right;
  };

  quantify(testCases, runs = 1e6) {
    const runsArr = Array.from({ length: runs });
    const solStart = performance.now();
    runsArr.map((run, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.method(input[0], input[1]));
        else this.method(input[0], input[1]);
      });
    });
    console.log(
      `Runtime for solution: ${(performance.now() - solStart) / 1000}\n`
    );

    const refStart = performance.now();
    runsArr.map((run, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.reference(input[0], input[1]));
        else this.reference(input[0], input[1]);
      });
    });
    console.log(
      `Runtime for reference: ${(performance.now() - refStart) / 1000}`
    );
  }
}

const test = new Solution();
const testCases = [
  [
    new TreeNode(1, new TreeNode(2), new TreeNode(3)),
    new TreeNode(1, new TreeNode(2), new TreeNode(3)),
  ],
  [new TreeNode(1, new TreeNode(2)), new TreeNode(1, null, new TreeNode(2))],
  [
    new TreeNode(1, new TreeNode(2), new TreeNode(1)),
    new TreeNode(1, new TreeNode(1), new TreeNode(2)),
  ],
];
test.quantify(testCases);
